packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -1027,6 +1033,32 @@ class RenderParagraph extends RenderBox
     return _textPainter.getWordBoundary(position);
   }
 
+  TextRange _getLineAtOffset(TextPosition position) => _textPainter.getLineBoundary(position);
+
+  TextPosition _getTextPositionAbove(TextPosition position) {
+    // The caret offset gives a location in the upper left hand corner of
+    // the caret so the middle of the line above is a half line above that
+    // point and the line below is 1.5 lines below that point.
```

Comment:
Super nit here and below:  You mention the caret here, but there's nothing in this method directly related to the caret (I see that you're using getOffsetForCaret in _getTextPositionVertical, but it's not obvious looking only at this method why you mention the caret).

Also similarly, it's not clear in this method why you're trying to find a point in the middle of the line (I guess you just need a point that's going to give you what you want from getPositionForOffset).